### DISCUSSION TITLE
abseil库链接问题

### DISCUSSION THREAD

**Discussion by Wkh3** [xmake-io/xmake#3648] [Q&A]
使用xmake导入abseil出现链接错误,示例代码如下:
![image](https://user-images.githubusercontent.com/59428912/232791527-29d0d27d-2342-410e-b1bd-42af4ba2d116.png)

我尝试用cmake链接abseil::log可以正常编译
![image](https://user-images.githubusercontent.com/59428912/232791998-560bc8e9-4f0e-4211-ad7a-237e92a244f6.png)

翻了一下abseil的cmake module,发现这里的absl_log生成的是一个interface target,不会导出静态库,不知道xmake能否识别?

**Comment by waruqi** [comment]
xmake -v 看下实际 link 了哪些库，排查下连接顺序，并用 nm 链接库是否存在对应符号

如果有问题，可以提个 pr 过来修复下包

**Comment by Wkh3** [comment]
我尝试在20230125.2版本手动链接absl log内部target,添加测试代码后可以成功build
![image](https://user-images.githubusercontent.com/59428912/232958873-cfc390a9-461c-4b8c-a147-28aba9cb8699.png)
但我翻看了一下其他cmake包,似乎都不需要手动link,有没有其他方式可以让xmake自动识别呢

**Reply by waruqi** [reply]
> 但我翻看了一下其他cmake包,似乎都不需要手动link

如果走动态库不需要，得先确认下。。xmake 默认安装走的是 静态库。。

**Reply by Wkh3** [reply]
我发先之前有大量手动link,这可能影响了xmake自身的link处理
![image](https://user-images.githubusercontent.com/59428912/232968440-da50c7c1-883d-4fe3-ae0b-23c685d00ad1.png)
将这段手动link删除后后可以成功work
![image](https://user-images.githubusercontent.com/59428912/232968749-3dffaa88-a4ef-4953-a615-08f5a6161441.png)

**Reply by waruqi** [reply]
> 我发先之前有大量手动link,这可能影响了xmake自身的link处理 ![image](https://user-images.githubusercontent.com/59428912/232968440-da50c7c1-883d-4fe3-ae0b-23c685d00ad1.png) 将这段手动link删除后后可以成功work ![image](https://user-images.githubusercontent.com/59428912/232968749-3dffaa88-a4ef-4953-a615-08f5a6161441.png)

就是要手动配置 links 的，你删了，只是让 xmake 默认走 lib 目录下自动扫描 .a 库，这样有一个问题是，扫出来加的 links 顺序是随机的，未定义行为，有可能是正好可以，有可能换个系统就不行。。

而 abseil 子库太多，里面很多有顺序依赖的，所以目前写死调整链接顺序。。

如果部分子库链接顺序还是不对，可以提 pr 过来微调那段 add_links ，但不能删。。

**Reply by waruqi** [reply]
你删了可以，说明还是链接顺序问题，确认下顺序，走 pr 调整下就行了。。

**Reply by Wkh3** [reply]
cmake似乎会根据库deps关系,自动调整链接顺序,上面的示例cmake只需要链接absl::log即可,xmake没有这样的处理嘛,对于这么庞大的库,人工处理link顺序很容易出错,一旦库更新又需要调整

**Reply by waruqi** [reply]
> cmake似乎会根据库deps关系,自动调整链接顺序,上面的示例cmake只需要链接absl::log即可,xmake没有这样的处理嘛,对于这么庞大的库,人工处理link顺序很容易出错,一旦库更新又需要调整

那是因为 abseil 也是 cmake 维护的脚本，自己当然很容易知道自己的依赖关系。。这边只能通过 add_links 。走自动分析反而不可靠。。xmake 还要去解析 cmakelists

**Reply by Wkh3** [reply]
cmake会将库的依赖关系写在pkg-config中的吧，xmake通过pkg-config获取对应库的依赖不行嘛
